mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMavericksViewModel.kt

```diff
@@ -164,6 +169,84 @@ abstract class BaseMavericksViewModel<S : MvRxState>(
+     *
+     * @param dispatcher The coroutine dispatcher that the coroutine will run on. Defaults to [Dispatchers.Main.immediate].
+     * @param retainValue A state property that, when set, will be called to retrieve an optional existing data value that will be retained across
+     *                    subsequent Loading and Fail states. This is useful if you want to display the previously succcessful data when
+     *                    refreshing.
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and it likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    fun <T : Any?> (suspend () -> T).execute(
+        dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+        retainValue: KProperty1<S, Async<T>>? = null,
+        reducer: S.(Async<T>) -> S
+    ): Job {
+        setState { reducer(Loading(value = retainValue?.get(this)?.invoke())) }
+        return viewModelScope.launch(dispatcher) {
+            try {
+                val result = invoke()
+                setState { reducer(Success(result)) }
+            } catch (e: CancellationException) {
+                @Suppress("RethrowCaughtException")
+                throw e
+            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                setState { reducer(Fail(e, value = retainValue?.get(this)?.invoke())) }
+            }
+        }
+    }
+
+    /**
+     * Collect a Flow and wrap its progression with [Async] property reduced to the global state.
+     *
+     * @param dispatcher The coroutine dispatcher that the coroutine will run on. Defaults to [Dispatchers.Main.immediate].
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and it likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    fun <T> Flow<T>.execute(
+        dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+        reducer: S.(Async<T>) -> S
+    ): Job {
+        setState { reducer(Loading<T>()) }
+        return onEach {
+            setState { reducer(Success(it)) }
+        }.launchIn(viewModelScope + dispatcher)
+    }
+
+    /**
+     * Collect a Flow and update state each time it emits a value. This is functionally the same as wrapping onEach with a setState call.
+     *
+     * @param dispatcher The coroutine dispatcher that the coroutine will run on. Defaults to [Dispatchers.Main.immediate].
+     * @param reducer A reducer that is applied to the current state and should return the new state. Because the state is the receiver
+     *                and it likely a data class, an implementation may look like: `{ copy(response = it) }`.
+     */
+    fun <T> Flow<T>.setOnEach(
```

Comment:
we have a similar extension at airbnb named `executeWithoutAsync`. kind of an ugly name, but I like that it 1) is discoverable since people are used to the "execute" function and autocomplete helps find this. 2) helps to explain the difference with "execute"

What do you think of some other name for this and starts with execute to help get those benefits?

Reply:
@elihart I see your point but I do think that `setOnEach` is more accurate since it really has nothing to do with execute at all.